add bootstrapping secrets page (#3476)
weaveworks/weave-gitops: website/docs/secrets/bootstraping-secrets.mdx

title: Bootstrapping Secrets
hide_title: true
---
import AlphaWarning from "../_components/_alpha_warning.mdx";

# Overview

When accessing protected resources there is a need for a client to authenticate before
the access is granted and the resource is consumed. For authentication, a client presents
credentials that are either created manually or available through infrastructure. A common scenario
is to have a secrets store.

Weave Gitops allows you to provision the secret management infrastructure as part of [its capabilities](./setup-secrets-operator.mdx).
However, in order to provision, as any other application that has secrets, we need to create the secret needed for installing it.
This is known as a chicken-egg scenario that get addressed by providing an initial secret. This secret we call it
bootstrapping secret.

> Bootstrapping secrets will be, not only present while provision your secrets management solution, but in

Reviewer: I still find this whole paragrap a bit complex and hard to follow. If we are trying to tell the user that the `bootstrapping secrets` functionally is beneficial for other things as well, then we can say something like:
`Bootstrapping secrets comes in handy, not only while provisioning your secrets management solution, but also in any platform provisioning task where the existence of the secret is a prerequisite.  Another common example could be provisioning
platform capabilities via [profiles](../../../cluster-management/getting-started/#profiles-and-clusters)
that are stored in [private repositories](https://fluxcd.io/flux/guides/helmreleases/#helm-repository-authentication-with-credentials).`

but also in any platform provisioning task where the existence of the secret is a prerequisite.
Another common example could be provisioning platform capabilities via [profiles](../../../cluster-management/getting-started/#profiles-and-clusters)
that are stored in [private repositories](https://fluxcd.io/flux/guides/helmreleases/#helm-repository-authentication-with-credentials).

Weave Gitops provides [SecretSync](#secretsync) as a solution to managing your bootstrapping secrets.

## SecretSync

<AlphaWarning/>

> `SecretSync` is a [Kubernetes Customer Resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)

Reviewer: ```suggestion
`SecretSync` is a [Kubernetes Custom Resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
```

organisation that already leverages terraform for cluster management infrastructure and want to created integrated workflows.

### Before start

- You have [TF-controller installed](../../terraform/get-started) in your management cluster
- You have some bootstrap secrets in the management cluster.
- You have a leaf cluster that you want to bootstrap secrets to.
- TF-controller has permissions to execute against the leaf cluster.

For example, in the context of [eks](../../terraform/aws-eks.mdx) you could leverage IRSA

```yaml
---
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: tf-controller
  namespace: flux-system
spec:
...
  values:
    ...
    runner:
      serviceAccount:
        annotations:
          eks.amazonaws.com/role-arn: "arn:aws:iam::894516026745:role/leaf-tf-controller" # TODO: replace with your role
```

//TODO limit the role

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: tf-runner
roleRef:
  kind: ClusterRole
  name: tf-runner
  apiGroup: rbac.authorization.k8s.io
subjects:
  - kind: ServiceAccount
    name: tf-runner
    namespace: flux-system
```

### Bootstrap secrets

> //TODO how to distribute the module currently lives in wge

Author: to par with @waleedhammam on how to distribute the modules

# Overview

When accessing to protected resources there is a need for a client to authenticate before
the access is granted and the resource is consumed. For the authentication, a client presents
credentials that are either created manually or available through infrastructure a common scenario
is a secrets store.

Weave gitops allows you to provision the secret management infrastructure as part of [its capabilities](./setup-secrets-operator.mdx).
However, in order to provision the secret infrastructure, as any other application that has secrets, we

> need to make the secret it needs to install it.

Author: applied, thanks you 🙏

This type of initial secret is what we call it bootstrapping secret and this guide gives you
with the different alternatives that weave gitops gives you to achieve the task.

> Bootstrapping secrets will be not only present while provision your secrets managmenet solution but in

Author: I reworded it yesterday before reading this comment. let me know whether the current version improves understanding or whether we should make it simpler.

any platform provisioning task that comes before you have avaialble before the secrets management solution
is availabe. Another common example is provisioning platform capabilities itself via
https://docs.gitops.weave.works/docs/cluster-management/getting-started/#profiles-and-clusters
that are stored in private repositories. For example https://fluxcd.io/flux/guides/helmreleases/#helm-repository-authentication-with-credentials

To address this problem, weave gitops provides the following alternatives:

- SecretSync CRD: to leverage weave gitops for secrets bootstrapping.

> - SecretSync Terraform Module: to leverage terraform and t for secrets bootstrapping.

Author: no longer tf docs

## Using `SecretSync` Custom Resource

`SecretSync` is a [Kubernetes Customer Resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
that provides semantics to sync [Kuberentes Secrets](https://kubernetes.io/docs/concepts/configuration/secret/) from management cluster to leaf clusters.

An example could be seen below:

```yaml
apiVersion: capi.weave.works/v1alpha1
kind: SecretSync
metadata:
  name: my-dev-secret-syncer
  namespace: default
spec:
  clusterSelector:
    matchLabels:
      environment: dev
  secretRef:
    name: my-dev-secret
  targetNamespace: my-namespace
```
Where you could find two main configuration sections:

1) Select the secret to sync:

```yaml
  secretRef:
    name: my-dev-secret
```

2) Select the [GitopsClusters](https://docs.gitops.weave.works/docs/cluster-management/managing-existing-clusters/)
,that the secret will be synced to, via labels:

```yaml
  clusterSelector:
    matchLabels:
      environment: dev
```

More info about the CRD spec [here](./spec/v1alpha1/secretSync.mdx)

### Working with SecretSync

> #### Before start

Reviewer: We can use `Pre-requisites` instead of `Before start` to be consistent with other pages.

Author: done ✔️

> 1. You are using [Weave Gitops Enterprise version > v0.19.0](https://docs.gitops.weave.works/docs/releases/#v0190)

Reviewer: A bit of nitpicking from my side, in this section you use numbers, and in the other `before start` section you use bullet points. Would be nice to pick one and go with it everything.

Author: we should have now only one (given terraform is not yet in). happy to put bullet points if you consider makes more sense here.

Reviewer: I dont have a preference. Whatever makes more sense to you/is more consistent with the website.

2. You are running cluster-bootstrap-controller > v0.5.0
2. You have a secret that you want to sync from the management cluster.

> 3. You have a set of GitopsClusters that you want to sync to.

Author: added thanks you

#### Syncing secrets via SecretSync

1. Create SecretSync manifests using the secret for example

```yaml
apiVersion: capi.weave.works/v1alpha1
kind: SecretSync
metadata:
  name: my-dev-secret-syncer
  namespace: default
spec:
  clusterSelector:
    matchLabels:
      environment: dev
  secretRef:
    name: my-dev-secret
  targetNamespace: my-namespace
```

2. Check-in to your configuration repo within your management cluster

3. Create a PR, review and merge

4. See the progress

```bash
kubectl get secretsyncs.capi.weave.works my-dev-secret-syncer -o yaml

apiVersion: capi.weave.works/v1alpha1
kind: SecretSync
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"capi.weave.works/v1alpha1","kind":"SecretSync","metadata":{"annotations":{},"name":"my-dev-secret-syncer","namespace":"default"},"spec":{"clusterSelector":{"matchLabels":{"environment":"dev"}},"secretRef":{"name":"my-dev-secret"},"targetNamespace":"my-namespace"}}
  creationTimestamp: "2023-03-28T16:44:37Z"
  generation: 1
  name: my-dev-secret-syncer
  namespace: default
  resourceVersion: "4980"
  uid: f6670578-4403-46f8-a9b5-de8c80559c2b
spec:

>   clusterSelector:

Author: to review with @ahsayde and bootstrap controller v0.5.0

I cannot set the resource being reconciled

    matchLabels:
      environment: dev
  secretRef:
    name: my-dev-secret
  targetNamespace: my-namespace
```
5. See the secret created in your leaf cluster

> CLI

Reviewer: Just adding a comment here as a reminder to add the CLI example

Author: done added yesterday

> UI

Reviewer: we dont show kubernetes secrets in the UI

Author: i guess it was more to show secretsync status from the UI

## Using Terraform and `TF-controller`

You could use terraform and tf-controller to provision the secrets. This scenario would be more suitable for an

> organisation that already leverages terraform for cluster management infrastructure and want to created integrated workflows.

Reviewer: ```suggestion
organisation that already leverages terraform for cluster management infrastructure and wants to create integrated workflows.
```

Author: terraform solution is not yet shipped so we could ignore

> ### Before start

Reviewer: We can use `Pre-requisites` instead of `Before start` to be consistent with other pages.

roleRef:
  kind: ClusterRole
  name: tf-runner
  apiGroup: rbac.authorization.k8s.io
subjects:
  - kind: ServiceAccount
    name: tf-runner
    namespace: flux-system
```

### Bootstrap secrets

//TODO how to distribute the module currently lives in wge
//probably just documentation at this stage
Check in your configuration repo Create a PR pointing to your terraform module for syncing secrets.

Adjust the configurationf your leaf cluster

Create the PR, review and merge

Once merged, see via tf-controller UI how the secrests has been created.

**Note (To be fixed)**

- you may have the following error

```bash
│ The "for_each" map includes keys derived from resource attributes that cannot be determined until apply, and so Terraform cannot determine the full set of keys that will
│ identify the instances of this resource.
│
│ When working with unknown values in for_each, it's better to define the map keys statically in your configuration and place apply-time results only in the map values.
│
│ Alternatively, you could use the -target planning option to first apply only the resources that the for_each value depends on, and then apply a second time to fully converge.
╵
```

> - This is because the flux-system have to be applied before flux installation

Reviewer: ```suggestion
- This is because the flux-system has to be applied before flux installation
```

Author: not yet included this piece

## Using `SecretSync` Custom Resource

`SecretSync` is a [Kubernetes Customer Resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
that provides semantics to sync [Kuberentes Secrets](https://kubernetes.io/docs/concepts/configuration/secret/) from management cluster to leaf clusters.

An example could be seen below:

```yaml
apiVersion: capi.weave.works/v1alpha1
kind: SecretSync
metadata:
  name: my-dev-secret-syncer
  namespace: default
spec:
  clusterSelector:
    matchLabels:
      environment: dev
  secretRef:
    name: my-dev-secret
  targetNamespace: my-namespace
```
Where you could find two main configuration sections:

1) Select the secret to sync:

```yaml
  secretRef:
    name: my-dev-secret
```

2) Select the [GitopsClusters](https://docs.gitops.weave.works/docs/cluster-management/managing-existing-clusters/)
,that the secret will be synced to, via labels:

```yaml
  clusterSelector:
    matchLabels:
      environment: dev
```

More info about the CRD spec [here](./spec/v1alpha1/secretSync.mdx)

### Working with SecretSync

#### Before start

> 1. You are using [Weave Gitops Enterprise version > XX](https://docs.gitops.weave.works/docs/releases/) with support for SecretsSync.

Author: to review with @ahsayde minimum version available

Author: ticket to release it https://github.com/weaveworks/weave-gitops-enterprise/issues/2500

Author: released and tested

2. You have a secret that you want to sync from the management cluster.
3. You have a set of GitopsClusters that you want to sync to.

#### Syncing secrets via SecretSync

1. Create SecretSync manifests using the secret for example

```yaml
```yaml
apiVersion: capi.weave.works/v1alpha1
kind: SecretSync
metadata:
  name: my-dev-secret-syncer
  namespace: default
spec:
  clusterSelector:
    matchLabels:
      environment: dev
  secretRef:
    name: my-dev-secret
  targetNamespace: my-namespace
```

2. Check-in to your configuration repo within your managmenet cluster

3. Create a PR, review and merge

4. See the progress via UI // CLI // TODO does it have support?

> CLI

Author: to review with @MostafaMegahid @ahsayde whether on this experience

UI

5. See the secret created in your leaf cluster

> CLI

Author: same here as before
